### DISCUSSION TITLE
Recommended flow for multiple error paths?

### DISCUSSION THREAD

**Discussion by derrickpelletier** [statelyai/xstate#1293] [General] [answered]
I'm implementing a machine that has a single idle state and then a series of 4 different invocations.

A couple of these invocations have multiple potential errors that can occur, and I'd like to reflect those errors distinctly in the UI when transitioning to the idle state. I have a working implementation, but I'm new to state machines so trying to get a sense of the best way to model this.

1. Should each of those known-error-states be _actual_ states? Like this: https://xstate.js.org/viz/?gist=73f0310b1cd09b183373fcce2f11c2cd. If so, whats the best way to target the known_error states via the invocation?

2. Should I just use actions to set the error and transition to a single idle state? Like this: https://xstate.js.org/viz/?gist=4b48fcbbdcc31e456bd286e017866c23. This means I likely need custom Error instances to handle the different errors in the UI where i interpret the machine and define the actions. But it also feels like the leanest implementation and leaves the details more up to the services than the machine.

3. Should I do the same as #2, but assign the error to context instead? https://xstate.js.org/viz/?gist=1a1d2af0b4b3f328601da5fe5952e312. If so, i feel like i'm doing redundant conditions. I.e. one _inside_ the invocation to determine the known_error, `throw`, then the condition inside the `onError`'s to determine if it was that same error and set the right transition.

Am I missing something obvious? 😇

**Comment by Andarist** [comment, marked as answer]
> Should each of those known-error-states be actual states?

It depends 😉 If you don't need to act differently when being in those error states then maybe it's not worth to separate them and only set a context value with appropriate error type.

> Like this: https://xstate.js.org/viz/?gist=73f0310b1cd09b183373fcce2f11c2cd. If so, whats the best way to target the known_error states via the invocation?

Not sure what do you mean - but my **guess** is that you want to put your error type into error event's payload and based on that select the known error state using conditional transitions for `onError`.

> Should I just use actions to set the error and transition to a single idle state? Like this: https://xstate.js.org/viz/?gist=4b48fcbbdcc31e456bd286e017866c23. This means I likely need custom Error instances to handle the different errors in the UI where i interpret the machine and define the actions. But it also feels like the leanest implementation and leaves the details more up to the services than the machine.

This sounds good to me, although I don't quite understand what do you mean by "I likely need custom Error instances to handle the different errors in the UI where i interpret the machine and define the actions".

> Should I do the same as #2, but assign the error to context instead? https://xstate.js.org/viz/?gist=1a1d2af0b4b3f328601da5fe5952e312. If so, i feel like i'm doing redundant conditions. I.e. one inside the invocation to determine the known_error, throw, then the condition inside the onError's to determine if it was that same error and set the right transition.

As mentioned in one of the previous parts of this answer - you can just set a value in your context based on the event's payload, so you should be able to avoid redundancy quite easily.

**Reply by derrickpelletier** [reply]
Thanks for the reply! Everything here makes sense.

If anything, this was validating. I actually chose the path in the second example.

What i meant by "I likely need custom Error instances to handle the different errors in the UI where i interpret the machine and define the actions" is more just an implementation detail of how to relate the UI to the state machine and the errors it's generating.

In some ways I was expecting an idiomatic style, but it seems that's just subjective 😀

**Comment by crishoj** [comment]
> put your error type into error event's payload and based on that select the known error state using conditional transitions for `onError`.

@Andarist Does XState allow multiple/conditional transitions for `onError`/`onDone`?
I've tried this, but got an error message along the lines of _Two error events with the same invocation and source are not allowed_.

**Reply by davidkpiano** [reply]
Can you share your code?

**Reply by crishoj** [reply]
What I initially tried was this:

```ts
invoke: {
    src: "SomeService",
    onDone: {target: "StateB"},
    onError: [
        {target: "StateC", actions: 'assignError', cond: 'someGuard'},
        {target: "StateD", actions: 'assignError'},
    ],
}
```

While it did seem to work, the editor at https://stately.ai did not render the conditional transition for `onError`, and IIRC it gave me an error message along the lines of _Two error events with the same invocation and source are not allowed._
